apis/metal3.io/v1alpha1/baremetalhost_validation.go

```diff
@@ -42,14 +55,93 @@ func (host *BareMetalHost) validateChanges(old *BareMetalHost) []error {
 	return errs
 }
 
-func validateRAID(r *RAIDConfig) error {
-	if r == nil {
+func validateRAID(s BareMetalHostSpec) error {
+	if s.RAID == nil {
 		return nil
 	}
 
-	if len(r.HardwareRAIDVolumes) > 0 && len(r.SoftwareRAIDVolumes) > 0 {
+	if len(s.RAID.HardwareRAIDVolumes) > 0 && len(s.RAID.SoftwareRAIDVolumes) > 0 {
 		return fmt.Errorf("hardwareRAIDVolumes and softwareRAIDVolumes can not be set at the same time")
 	}
 
+	bmcType, err := getBMCType(s.BMC.Address)
+	if err != nil {
+		return err
+	}
+
+	switch bmcType {
+	case "ilo5", "idrac", "irmc":
+		return nil
+	default:
+		return fmt.Errorf("currently the 'raid' field is only supported by ilo5/idrac/irmc")
+	}
+}
+
+func getBMCType(address string) (string, error) {
+	strs := strings.Split(address, "://")
+	if len(strs) != 2 {
+		return "", fmt.Errorf("the format of bmc.address: %s is wrong", address)
+	}
+
+	return strings.Split(strs[0], "+http")[0], nil
+}
+
+func validateBMC(b BMCDetails) error {
+	if b.Address == "" {
+		return nil
+	}
+
+	avaliableType := []string{
+		"ipmi", "idrac", "idrac-virtualmedia", "idrac-redfish",
+		"irmc", "ibmc", "ilo4", "ilo5", "ilo5-redfish",
+		"redfish", "redfish-virtualmedia", "libvirt",
+	}
+	m := make(map[string]bool)
+	for _, v := range avaliableType {
+		m[v] = true
+	}
+
+	bmcType, err := getBMCType(b.Address)
+	if err != nil {
+		return err
+	}
+	if _, ok := m[bmcType]; !ok {
+		return fmt.Errorf("unsupport BMC Type %s", bmcType)
+	}
 	return nil
 }
+
+func validateBootMACAddress(s BareMetalHostSpec) error {
+	if s.BMC.Address == "" {
+		return nil
+	}
+
+	bmcType, err := getBMCType(s.BMC.Address)
+	if err != nil {
+		return err
+	}
+
+	if bmcType == "libvirt" && s.BootMACAddress == "" {
+		return fmt.Errorf("BootMACAddress is required for libvirt VMs driven by vbmc")
+	}
+
+	return nil
+}
+
+func validateFirmware(s BareMetalHostSpec) error {
+	if s.Firmware == nil {
+		return nil
+	}
+
+	bmcType, err := getBMCType(s.BMC.Address)
+	if err != nil {
+		return err
+	}
+
+	switch bmcType {
+	case "ilo4", "ilo5", "idrac", "irmc":
+		return nil
+	default:
+		return fmt.Errorf("currently the firmware field is only supported by ilo4/ilo5/irmc/idrac")
+	}
+}
```

Comment:
I agree with duplication is problematic. However, as we discussed previously, it is not easy currently. 
There are two ways to achieve this and prevent duplication;
- kubebuilder will support moving validations under `pkg` folder.
- or as @dtantsur stated in here https://github.com/openshift-metal3/terraform-provider-ironic/pull/53#issuecomment-925574411, these things are moved to metal3 agnostic module and safely used in here.

Reply:
And seems that some shared code (validation helpers) are needed to be separated into isolated (by import deps) package to keep that isolated code with minimum dependencies, that will avoid import loops in future.

Something like `common` package or `shared` package, that does not import anything except golang core libs and possibly `api` package

Reply:
I think it's a good approach in general @s3rj1k , and moreover I would suggest a slightly different way to face it. The basic idea is to have a sort of validation framework reflection based, both in order to break the dependencies where needed and also to have a more structured way to contribute new validation rules.
In the past I've used the [go-playground validator](https://github.com/go-playground/validator) for that: it essentially allows to define and apply validations rules in different ways (mostly by annotating fields)' it's shipped with a number of prebuilt validations rules but it's also pretty easy to define new one.
For basic checks (ie `BootMACAddress` is not empty) its application it's really straitghforward. But I think that whenever we have an interface available it could be also adopted for more complex scenarios, reducing a lot the dependency issue. Another option on the plate.

Reply:
Something like https://github.com/asaskevich/govalidator with their standalone functions, no need to add struct tags to trigger struct validation, this validation specific tags can become an issue

Reply:
I updated this PR to call the operator method for validation of RAID and BIOS.
Added validation method for BootMACAddress in webhook. PTAL, thanks!